import os
configfile: "../config/config.yaml"

# base is a folder the reads that are stored as reads/SAMPLE/SAMPLE_{1-2}.fq.gz
base = config["readfolder"]

ids = os.listdir(base + "reads/")

want_all = []
want_all.append(expand(base + "nohuman/{sample}_noMatch_1.fq.gz", sample = ids))
want_all.append(expand(base + "AR/{sample}.bam", sample=ids))
want_all.append(expand(base + "nreads/{sample}.nreads", sample=ids))
want_all.append(base + "results/All.nreads")
want_all.append(expand(base + "AR/{sample}.bam.bai", sample=ids))
want_all.append(expand(base + "AR/{sample}.bam.stat", sample=ids))
want_all.append(expand(base + "AR/{sample}.RPKM", sample=ids))
want_all.append(base + "results/All.RPKM")
want_all.append(expand(base + "AR_red/{sample}.bgcov", sample=ids))
want_all.append(expand(base + "AR_red/{sample}.bgcov.wid", sample=ids))
want_all.append(expand(base + "AR_red/99_filt/{sample}.bgcov.wid.meg99", sample=ids))
want_all.append(expand(base + "AR_red/99_filt/{sample}.bgcov.wid.gene99", sample=ids))
want_all.append(expand(base + "AR_red/99_filt/{sample}.bgcov.wid.gene99", sample=ids))
want_all.append(expand(base + "AR_red/99_filt/{sample}.bgcov.wid.gene99.adiv", sample=ids))
want_all.append(base + "results/All_red.adiv")
want_all.append(expand(base + "taxa/{sample}.mph_report", sample=ids))
want_all.append(expand(base+"taxa/{sample}.RPM_mph", sample=ids))
want_all.append(base + "results/All.RPM_mph")

rule all:
    input:
        want_all,

# cutting adapters. files will be removed after obtaining "clean" reads
rule cutadapt:
    input:
        r1=base + "reads/{sample}/{sample}_1.fq.gz",
        r2=base + "reads/{sample}/{sample}_2.fq.gz",
    output:
        r1=temp(base + "cutadapt/{sample}_cut.1.fq.gz"),
        r2=temp(base + "cutadapt/{sample}_cut.2.fq.gz"),
        info=temp(base + "cutadapt/{sample}_cut.info"),
    conda:
        "envs/arsnake.yml"
    log:
        base + "logs/cutadapt/{sample}.log",
    shell:
        "cutadapt -a AAGTCGGAGGCCAAGCGGTCTTAGGAAGACAA -A AAGTCGGATCCTAGCCATGTCGTTCTGTGAG --minimum-length 1 --info-file {output.info} -o {output.r1} -p {output.r2}  {input.r1} {input.r2} \
  > {log} 2>&1;"

# trimming. files  will be removed after obtaining "clean" reads
rule trim:
    input:
        r1=base + "reads/{sample}/{sample}_1.fq.gz",
        r2=base + "reads/{sample}/{sample}_2.fq.gz",
    output:
        r1=base + "trimmed/{sample}_trim.1.fq.gz",
        r2=base + "trimmed/{sample}_trim.2.fq.gz",
        r1_unp=temp(base + "trimmed/{sample}_trim_unp.1.fq.gz"),
        r2_unp=temp(base + "trimmed/{sample}_trim_unp.2.fq.gz"),
    conda:
        "envs/arsnake.yml"
    log:
        base + "logs/trimmed/{sample}.log",
    shell:
        "trimmomatic PE -phred33 {input.r1} {input.r2} {output.r1} {output.r1_unp} {output.r2} {output.r2_unp} LEADING:25 TRAILING:25 SLIDINGWINDOW:4:15 MINLEN:50 &> {log}"


# locating biobloom filter files

# filtering human DNA and obtaining clean reads
rule biobloom:
    input:
        r1=base + "trimmed/{sample}_trim.1.fq.gz",
        r2=base + "trimmed/{sample}_trim.2.fq.gz",
    output:
        summary=base + "nohuman/{sample}_summary.tsv",
        r1_noMatch=base + "nohuman/{sample}_noMatch_1.fq.gz",
        r2_noMatch=base + "nohuman/{sample}_noMatch_2.fq.gz",
    params:
        dirr=base + "nohuman/{sample}",
        hg19_masked_bf = config["human_bf"],
        hg19_masked_txt = config["human_txt"],
    shell:
        """
        biobloomcategorizer --fq  -p  {params.dirr} -f {params.hg19_masked_bf} -e {input.r1} {input.r2} --gz_output &> {output.summary}
        """


# counting "clean" filtered reads
rule count_reads:
    input:
        base + "nohuman/{sample}_noMatch_1.fq.gz",
    output:
        base + "nreads/{sample}.nreads",
    shell:
        "echo $(echo $(zcat {input} | wc -l)/4 | bc) $(basename {input} _noMatch_1.fq.gz) > {output}"

# create nreads table for all samples
rule nreads_all:
    input:
        expand(base + "nreads/{sample}.nreads", sample=ids),
    output:
        base + "results/All.nreads",
    shell:
        "cat {input} > {output}"

# mapping "clean" reads to AR gene database
# BASIC mapping mode
rule map_AR:
    input:
        r1=base + "nohuman/{sample}_noMatch_1.fq.gz",
        r2=base + "nohuman/{sample}_noMatch_2.fq.gz",
    params:
        db=config["AR"],
    output:
        base + "AR/{sample}.bam",
    conda:
        "envs/arsnake.yml"
    log:
        base + "logs/AR/{sample}.mapAR.log",
    shell:
        "bowtie2 -x {params.db} -1 {input.r1} -2 {input.r2} -k 1 2> {log} | samtools view -b -q 30 | samtools sort -o {output} "

# indexing .bam files
rule bam_index:
    input:
        base + "AR/{sample}.bam",
    output:
        base + "AR/{sample}.bam.bai",
    conda:
        "envs/arsnake.yml"
    log:
        base + "logs/samtools/{sample}.index.log",
    shell:
        "samtools index {input} &> {log}"


# get mapping stats that is needed for RPKM
rule mapping_stat:
    input:
        bam=base + "AR/{sample}.bam",
        index=base + "AR/{sample}.bam.bai",
    output:
        base + "AR/{sample}.bam.stat",
    log:
        #in case something goes wrong
        base + "logs/samtools/{sample}.ERRlog",
    conda:
        "envs/arsnake.yml"
    shell:
        "samtools idxstats {input.bam} > {output}"


# get RPKM for each gene
rule rpkm:
    input:
        stat=base + "AR/{sample}.bam.stat",
        nrd=base + "nreads/{sample}.nreads",
    output:
        base + "AR/{sample}.RPKM",
    log:
        #in case something goes wrong
        base + "logs/RPKM/{sample}.ERRlog",
    shell:
        "python3 scripts/rpkm.py {input.nrd} {input.stat} > {output} 2> {log}"


# create RPKM table for all samples
rule rpkm_all:
    input:
        expand(base + "AR/{sample}.RPKM", sample=ids),
    output:
        base + "results/All.RPKM",
    shell:
        "cat {input} > {output}"

# REDUNDANT mapping mode
rule map_AR_red:
    input:
        r1=base + "nohuman/{sample}_noMatch_1.fq.gz",
        r2=base + "nohuman/{sample}_noMatch_2.fq.gz",
    params:
        db=config["AR"],
    output:
        temp(base + "AR_red/{sample}.bam"),
    log:
        base + "logs/AR_red/{sample}.mapAR.log",
    shell:
        "bowtie2 -x {params.db} -1 {input.r1} -2 {input.r2} -a 2> {log} | samtools view -b -q 30 | samtools sort -o {output}"


# get info on how many nucleotides are covered for each gene
rule bgcov:
    input:
        base + "AR_red/{sample}.bam",
    output:
        base + "AR_red/{sample}.bgcov",
    conda:
        "envs/arsnake.yml"
    log:
        #in case something goes wrong
        base + "logs/wid/{sample}.bgcov.ERRlog",
    shell:
        "bedtools genomecov -ibam {input} > {output} 2> {log}"


# get % of covered bases for each gene variant
rule wid:
    input:
        base + "AR_red/{sample}.bgcov",
    output:
        base + "AR_red/{sample}.bgcov.wid",
    log:
        #in case something goes wrong
        base + "logs/wid/{sample}.wid.ERRlog",
    shell:
        "python3 scripts/genomecov_to_wid.py {input} > {output} 2> {log}"


# get only sequences that are covered by 99% of their length
rule filter_99meg:
    input:
        base + "AR_red/{sample}.bgcov.wid",
    output:
        base + "AR_red/99_filt/{sample}.bgcov.wid.meg99",
    shell:
        "cat {input} | awk '{{if ($2 > 98) print}}' > {output}"


# get only genes that have at least one gene variant covered by 99% of its length
rule filter_99gene:
    input:
        base + "AR_red/99_filt/{sample}.bgcov.wid.meg99",
    output:
        base + "AR_red/99_filt/{sample}.bgcov.wid.gene99",
    shell:
        """cat {input} | awk '{{split($1,a,"|") ; print a[5]}}' | sort -u > {output}"""


# get resistome richness as a number of genes that have al least
# one gene variant covered by 99% of its length
rule adiv_99:
    input:
        base + "AR_red/99_filt/{sample}.bgcov.wid.gene99",
    output:
        base + "AR_red/99_filt/{sample}.bgcov.wid.gene99.adiv",
    shell:
        "s=$(cat {input} | wc -l); echo $s {wildcards.sample} > {output}"

# get table with resistome  values for all samples
rule adiv_all:
    input:
        expand(base + "AR_red/99_filt/{sample}.bgcov.wid.gene99.adiv", sample=ids),
    output:
        base + "results/All_red.adiv",
    shell:
        "cat {input} > {output}"


# get table with sample counts for each fully covered gene
# (needed for identifying "core" genes)
# NOTE: this file comes empty in test data
rule forcore:
    input:
        expand(base + "AR_red/99_filt/{sample}.bgcov.wid.gene99", sample=ids),
    output:
        base + "results/All.ARsample_counts",
    shell:
        "cat {input} | sort | uniq -c | sort -nk1 > {output}"

# build taxonomy profiles with metaphlan
rule metaphlan:
    input:
        r1=base + "nohuman/{sample}_noMatch_1.fq.gz",
        r2=base + "nohuman/{sample}_noMatch_2.fq.gz",
    output:
        bto=temp(base + "taxa/{sample}.mph_bowtie2out.bz2"),
        r=base + "taxa/{sample}.mph_report",
    log:
        log=base + "logs/taxa/{sample}.mph.log",
    shell:
        "metaphlan {input.r1},{input.r2} --input_type fastq --bowtie2out {output.bto} --ignore_eukaryotes -t rel_ab_w_read_stats -o {output.r} 2> {log}"

# convert abundance values to .RPM 
rule mph_rpm:
        input:
                mph=base+"taxa/{sample}.mph_report",
                nreads=base+"nreads/{sample}.nreads"
        output:
                base+"taxa/{sample}.RPM_mph"
        shell:
                "python3 scripts/reads_per_taxa.py {input.mph} {input.nreads} > {output}"

# create a summary table with all .RPM values
# Note: eukaryotes are ignored
# Note: this file may be big, consider analyzing it by taxa levels
rule mph_rpm_all:
    input:
        expand(base+"taxa/{sample}.RPM_mph", sample=ids),
    output:
        base + "results/All.RPM_mph",
    shell:
        "cat {input} > {output}"

